File: src/lib/api/import-tasks.ts
/**
 * # Import Tasks API Client
 *
 * Pure API layer for import task operations.
 *
 * ## Overview
 *
 * Encapsulates all Tauri `invoke` calls for import task persistence.
 * This module is stateless and has no side effects - it simply wraps
 * the backend commands with TypeScript types.
 *
 * ## Backend Commands Mapping
 *
 * | Function            | Backend Command                          |
 * |---------------------|------------------------------------------|
 * | listImportTasks     | list_import_tasks                        |
 * | listRecentTasks     | list_recent_import_tasks                 |
 * | getImportTask       | get_import_task                          |
 * | createImportTask    | create_import_task                       |
 * | updateTaskStatus    | update_import_task_status                |
 * | completeTask        | complete_import_task                     |
 * | failTask            | fail_import_task                         |
 * | prepareTaskRetry    | prepare_task_retry                       |
 * | deleteImportTask    | delete_import_task                       |
 * | cleanupOldTasks     | cleanup_old_import_tasks                 |
 * | processSourceWithEvents | process_source_recursive_with_events_cmd |
 * | processSource       | process_source_recursive_cmd             |
 *
 * ## Usage
 *
 * ```typescript
 * import * as api from '$lib/api/import-tasks';
 *
 * // Create and process a task
 * await api.createImportTask(id, path, name, targetLibrary);
 * const result = await api.processSourceWithEvents(id, path, 5);
 * await api.completeTask(id, dest, files, size, type);
 * ```
 */
import { invoke } from '@tauri-apps/api/core';
import type { ApiResponse, RecursiveExtractResult } from '$lib/api/commands';
import type { PersistedImportTask, ImportStatus } from '$lib/stores/imports';

// =============================================================================
// TASK LISTING
// =============================================================================

/**
 * Lists all import tasks from the database.
 * @returns All persisted tasks, ordered by creation date descending.
 */
export async function listImportTasks(): Promise<ApiResponse<PersistedImportTask[]>> {
  return invoke<ApiResponse<PersistedImportTask[]>>('list_import_tasks');
}

/**
 * Lists recent import tasks within a time window.
 * @param days - Number of days to look back.
 * @returns Tasks created within the specified time window.
 */
export async function listRecentTasks(days: number): Promise<ApiResponse<PersistedImportTask[]>> {
  return invoke<ApiResponse<PersistedImportTask[]>>('list_recent_import_tasks', { days });
}

/**
 * Retrieves a single task by ID.
 * @param id - UUID of the task.
 * @returns The task if found, null otherwise.
 */
export async function getImportTask(id: string): Promise<ApiResponse<PersistedImportTask | null>> {
  return invoke<ApiResponse<PersistedImportTask | null>>('get_import_task', { id });
}

// =============================================================================
// TASK LIFECYCLE
// =============================================================================

/**
 * Creates a new import task in the database.
 * @param id - UUID for the new task.
 * @param sourcePath - Path to the source archive or directory.
 * @param name - Display name (usually the filename).
 * @param targetLibrary - Optional target DAZ library path.
 */
export async function createImportTask(
  id: string,
  sourcePath: string,
  name: string,
  targetLibrary: string | null
): Promise<void> {
  await invoke('create_import_task', { id, sourcePath, name, targetLibrary });
}

/**
 * Updates the status of a task.
 * @param id - UUID of the task.
 * @param status - New status value.
 */
export async function updateTaskStatus(id: string, status: ImportStatus): Promise<void> {
  await invoke('update_import_task_status', { id, status });
}

/**
 * Marks a task as successfully completed.
 *
 * Also moves source archives to trash if the setting is enabled.
 *
 * @param id - UUID of the task.
 * @param destination - Final destination path.
 * @param filesCount - Number of files extracted.
 * @param totalSize - Total size in bytes.
 * @param contentType - Detected content type (e.g., "Character", "Clothing").
 * @param sourceArchivePaths - Paths to move to trash if setting enabled.
 */
export async function completeTask(
  id: string,
  destination: string,
  filesCount: number,
  totalSize: number,
  contentType: string | null,
  sourceArchivePaths?: string[]
): Promise<void> {
  await invoke('complete_import_task', { 
    id, 
    destination, 
    filesCount, 
    totalSize, 
    contentType,
    sourceArchivePaths: sourceArchivePaths ?? null
  });
}

/**
 * Marks a task as failed with an error message.
 * @param id - UUID of the task.
 * @param error - Error message describing the failure.
 */
export async function failTask(id: string, error: string): Promise<void> {
  await invoke('fail_import_task', { id, error });
}

/**
 * Prepares a task for retry by resetting its status.
 * @param id - UUID of the task.
 * @returns True if the task was successfully prepared.
 */
export async function prepareTaskRetry(id: string): Promise<ApiResponse<boolean>> {
  return invoke<ApiResponse<boolean>>('prepare_task_retry', { id });
}

// =============================================================================
// TASK CLEANUP
// =============================================================================

/**
 * Deletes a task from the database.
 * @param id - UUID of the task to delete.
 */
export async function deleteImportTask(id: string): Promise<void> {
  await invoke('delete_import_task', { id });
}

/**
 * Cleans up old completed tasks.
 * @param days - Delete tasks older than this many days.
 * @returns Number of tasks deleted.
 */
export async function cleanupOldTasks(days: number): Promise<ApiResponse<number>> {
  return invoke<ApiResponse<number>>('cleanup_old_import_tasks', { days });
}

/**
 * Clears all completed tasks (done or error status) from the database.
 * @returns Number of tasks deleted.
 */
export async function clearCompletedTasks(): Promise<ApiResponse<number>> {
  return invoke<ApiResponse<number>>('clear_completed_import_tasks');
}

// =============================================================================
// PROCESSING
// =============================================================================

/**
 * Processes a source with recursive extraction and progress events.
 *
 * This is the main processing function that:
 * 1. Extracts archives recursively (up to maxDepth levels)
 * 2. Emits "import_step" events for UI progress
 * 3. Copies content to the DAZ library
 *
 * @param taskId - UUID of the task (for event correlation).
 * @param path - Path to the source archive or directory.
 * @param maxDepth - Maximum nesting depth for recursive extraction.
 * @returns Extraction result with statistics.
 */
export async function processSourceWithEvents(
  taskId: string,
  path: string,
  maxDepth: number
): Promise<ApiResponse<RecursiveExtractResult>> {
  return invoke<ApiResponse<RecursiveExtractResult>>(
    'process_source_recursive_with_events_cmd',
    { taskId, path, maxDepth }
  );
}

/**
 * Processes a source without progress events.
 * Simpler version for batch processing without UI feedback.
 * @param path - Path to the source archive or directory.
 * @returns Extraction result with statistics.
 */
export async function processSource(path: string): Promise<ApiResponse<RecursiveExtractResult>> {
  return invoke<ApiResponse<RecursiveExtractResult>>('process_source_recursive_cmd', {
    path,
    maxDepth: 5,
  });
}

File: src-tauri/src/commands/import_tasks.rs
//! # Import Task Commands
//!
//! Tauri commands for managing persisted import tasks.
//!
//! ## Overview
//!
//! Import tasks are persisted to SQLite to survive application restarts.
//! This enables:
//! - Resuming interrupted imports
//! - Retrying failed imports
//! - Viewing import history
//!
//! ## Task Lifecycle
//!
//! ```text
//! pending -> processing -> done
//!                      \-> error (retryable)
//!                      \-> interrupted (on app restart, retryable)
//! ```

use crate::config::settings::SETTINGS;
use crate::core::manifest;
use crate::core::trash;
use crate::db::{Database, NewProduct};
use crate::db::import_tasks::{ImportTaskStatus, ImportTasksRepository, PersistedImportTask};
use crate::error::{ApiResponse, AppError};
use chrono::{TimeZone, Utc};
use rusqlite::Connection;
use std::path::{Path, PathBuf};
use std::sync::Mutex;
use tauri::State;
use tracing::{info, warn};

use super::products::DbState;

// =============================================================================
// TAURI STATE
// =============================================================================

/// Thread-safe state wrapper for import tasks database access.
///
/// Holds a lazily-initialized SQLite connection for import task persistence.
pub struct ImportTasksState {
    repository: Mutex<Option<ImportTasksRepository>>,
}

impl ImportTasksState {
    /// Creates a new uninitialized state.
    pub fn new() -> Self {
        Self {
            repository: Mutex::new(None),
        }
    }

    /// Initializes the repository with the given database path.
    ///
    /// Must be called before any command that accesses the repository.
    /// Also marks any "processing" tasks as "interrupted" (app crashed/restarted).
    pub fn init(&self, db_path: &str) -> Result<(), String> {
        let connection =
            Connection::open(db_path).map_err(|e| format!("Failed to open database: {}", e))?;

        let repo = ImportTasksRepository::new(connection);

        // Create schema if needed
        repo.initialize()
            .map_err(|e| format!("Failed to initialize schema: {}", e))?;

        // Mark "processing" tasks as "interrupted" (app was restarted)
        if let Err(e) = repo.mark_interrupted() {
            tracing::warn!("Failed to mark interrupted tasks: {}", e);
        }

        *self.repository.lock().map_err(|_| "Repository mutex poisoned".to_string())? = Some(repo);
        Ok(())
    }

    /// Executes a function with access to the repository.
    fn with_repo<T, F>(&self, operation: F) -> Result<T, String>
    where
        F: FnOnce(&ImportTasksRepository) -> Result<T, crate::error::AppError>,
    {
        let guard = self.repository.lock().map_err(|_| "Lock error")?;
        let repo = guard.as_ref().ok_or("Repository not initialized")?;
        operation(repo).map_err(|e| e.to_string())
    }
}

// =============================================================================
// CRUD COMMANDS
// =============================================================================

/// Creates a new import task in the database.
#[tauri::command]
pub async fn create_import_task(
    state: State<'_, ImportTasksState>,
    id: String,
    source_path: String,
    name: String,
    target_library: Option<String>,
) -> Result<ApiResponse<bool>, String> {
    info!("Creating import task: {} - {}", id, name);

    state
        .with_repo(|repo| {
            let task = PersistedImportTask {
                id: id.clone(),
                source_path,
                name,
                status: ImportTaskStatus::Pending,
                destination: None,
                error_message: None,
                files_count: None,
                total_size: None,
                content_type: None,
                started_at: chrono::Utc::now().timestamp_millis(),
                completed_at: None,
                target_library,
            };
            repo.add_task(&task)?;
            Ok(true)
        })
        .map(ApiResponse::success)
        .or_else(|e| Ok(ApiResponse::error_msg("CREATE_ERROR", &e)))
}

/// Updates the status of an existing task.
#[tauri::command]
pub async fn update_import_task_status(
    state: State<'_, ImportTasksState>,
    id: String,
    status: String,
) -> Result<ApiResponse<bool>, String> {
    let parsed_status = match status.as_str() {
        "pending" => ImportTaskStatus::Pending,
        "processing" => ImportTaskStatus::Processing,
        "done" => ImportTaskStatus::Done,
        "error" => ImportTaskStatus::Error,
        "interrupted" => ImportTaskStatus::Interrupted,
        _ => {
            return Ok(ApiResponse::error_msg(
                "INVALID_STATUS",
                &format!("Unknown status: {}", status),
            ))
        }
    };

    state
        .with_repo(|repo| {
            repo.update_status(&id, parsed_status)?;
            Ok(true)
        })
        .map(ApiResponse::success)
        .or_else(|e| Ok(ApiResponse::error_msg("UPDATE_ERROR", &e)))
}

/// Marks a task as successfully completed.
///
/// Also moves source archives to trash if the `trash_archives_after_import`
/// setting is enabled.
#[tauri::command]
pub async fn complete_import_task(
    state: State<'_, ImportTasksState>,
    products_db_state: State<'_, DbState>,
    id: String,
    destination: String,
    files_count: i64,
    total_size: i64,
    content_type: Option<String>,
    source_archive_paths: Option<Vec<String>>,
) -> Result<ApiResponse<bool>, String> {
    info!("Completing import task: {}", id);

    // Mark the task as completed in database
    let result = state.with_repo(|repo| {
        repo.set_result(
            &id,
            &destination,
            files_count,
            total_size,
            content_type.as_deref(),
        )?;
        Ok(true)
    });

    // Best-effort: create a product record from the completed task
    if result.is_ok() {
        match state.with_repo(|repo| repo.get_task(&id)) {
            Ok(Some(task)) => {
                if let Err(e) = upsert_product_from_task(&products_db_state, &task) {
                    warn!("Failed to create product for task {}: {}", id, e);
                }
            }
            Ok(None) => warn!("Cannot create product: task not found after completion: {}", id),
            Err(e) => warn!("Cannot create product: failed to fetch task {}: {}", id, e),
        }
    }

    // If import succeeded and trash setting is enabled, move archives to trash
    if result.is_ok() {
        let should_trash = SETTINGS
            .read()
            .map(|s| s.trash_archives_after_import)
            .unwrap_or(false);

        if should_trash {
            if let Some(archive_paths) = source_archive_paths {
                trash_source_archives(&archive_paths, &destination);
            }
        }
    }

    result
        .map(ApiResponse::success)
        .or_else(|e| Ok(ApiResponse::error_msg("COMPLETE_ERROR", &e)))
}

/// Creates a product record from a completed import task.
/// Returns `Ok(true)` if created, `Ok(false)` if skipped (already exists or missing data).
///
/// Also parses Manifest.dsx/Supplement.dsx from the destination to:
/// - Enrich product metadata (global_id, product name)
/// - Store per-file inventory in `product_files` table
fn upsert_product_from_task(
    products_db_state: &State<'_, DbState>,
    task: &PersistedImportTask,
) -> Result<bool, AppError> {
    let destination = match &task.destination {
        Some(d) => d,
        None => return Ok(false),
    };

    // Optional installed date from task completion timestamp
    let installed_at = task
        .completed_at
        .and_then(|ts| Utc.timestamp_millis_opt(ts).single())
        .map(|dt| dt.to_rfc3339());

    let mut new_product = NewProduct::new(&task.name, destination)
        .with_import_task_id(task.id.clone())
        .with_stats(task.files_count.unwrap_or(0), task.total_size.unwrap_or(0));

    if let Some(installed_at) = installed_at {
        new_product = new_product.with_installed_at(installed_at);
    }

    // Link source archive if the source is an archive file
    if trash::is_archive_file(Path::new(&task.source_path)) {
        new_product = new_product.with_source(task.source_path.clone());
    }

    if let Some(content_type) = &task.content_type {
        new_product = new_product.with_content_type(content_type.clone());
    }

    // Parse Manifest.dsx / Supplement.dsx from destination library
    let product_manifest = manifest::parse_product_manifests(Path::new(destination))
        .unwrap_or_else(|e| {
            warn!("Failed to parse manifests in {}: {}", destination, e);
            manifest::ProductManifest::default()
        });

    // Enrich product with manifest metadata
    if let Some(global_id) = &product_manifest.global_id {
        new_product = new_product.with_global_id(global_id.clone());
    }
    // Use Supplement product name if it's more descriptive than the archive filename
    if let Some(product_name) = &product_manifest.product_name {
        if !product_name.is_empty() {
            new_product.name = product_name.clone();
        }
    }

    let guard = products_db_state
        .0
        .lock()
        .map_err(|_| AppError::Internal("Database mutex poisoned".to_string()))?;

    let db: &Database = guard
        .as_ref()
        .ok_or_else(|| AppError::Internal("Database not initialized".to_string()))?;

    // Idempotency: skip if we already created a product for this task.
    if db.get_product_by_import_task_id(&task.id)?.is_some() {
        return Ok(false);
    }

    let product_id = db.add_product(&new_product)?;

    // Store per-file inventory from manifest
    if !product_manifest.files.is_empty() {
        let file_entries: Vec<(String, String)> = product_manifest
            .files
            .iter()
            .map(|f| (f.relative_path.clone(), f.target.clone()))
            .collect();

        db.with_connection(|conn| {
            crate::db::product_files::insert_product_files_batch(conn, product_id, &file_entries)
        })
        .unwrap_or_else(|e| {
            warn!("Failed to store product files for {}: {}", task.name, e);
            0
        });

        info!(
            "Stored {} manifest files for product {} (id={})",
            file_entries.len(),
            new_product.name,
            product_id
        );
    }

    Ok(true)
}

/// Moves source archives to trash after successful import.
fn trash_source_archives(archive_paths: &[String], destination: &str) {
    for source in archive_paths {
        let source_path = Path::new(source);

        // Only trash actual archives (not folders)
        if !trash::is_archive_file(source_path) {
            continue;
        }

        // Safety: don't trash if destination is inside source
        let dest_path = Path::new(destination);
        if source_path == dest_path || destination.starts_with(source) {
            continue;
        }

        // Skip if archive no longer exists (already removed during extraction)
        if !source_path.exists() {
            info!("Archive already removed, skipping trash: {}", source);
            continue;
        }

        match trash::move_to_trash(source_path) {
            Ok(true) => {
                info!("Archive moved to trash after successful import: {}", source);
            }
            Ok(false) => {
                info!("Archive not found for trash: {}", source);
            }
            Err(e) => {
                // Trash failure is a warning only, doesn't affect import success
                warn!("Failed to move archive to trash: {} - {}", source, e);
            }
        }
    }
}

/// Marks a task as failed with an error message.
#[tauri::command]
pub async fn fail_import_task(
    state: State<'_, ImportTasksState>,
    id: String,
    error: String,
) -> Result<ApiResponse<bool>, String> {
    info!("Failing import task: {} - {}", id, error);

    state
        .with_repo(|repo| {
            repo.set_error(&id, &error)?;
            Ok(true)
        })
        .map(ApiResponse::success)
        .or_else(|e| Ok(ApiResponse::error_msg("FAIL_ERROR", &e)))
}

// =============================================================================
// QUERY COMMANDS
// =============================================================================

/// Retrieves a single task by ID.
#[tauri::command]
pub async fn get_import_task(
    state: State<'_, ImportTasksState>,
    id: String,
) -> Result<ApiResponse<Option<PersistedImportTask>>, String> {
    state
        .with_repo(|repo| repo.get_task(&id))
        .map(ApiResponse::success)
        .or_else(|e| Ok(ApiResponse::error_msg("GET_ERROR", &e)))
}

/// Lists all import tasks.
#[tauri::command]
pub async fn list_import_tasks(
    state: State<'_, ImportTasksState>,
) -> Result<ApiResponse<Vec<PersistedImportTask>>, String> {
    state
        .with_repo(|repo| repo.list_tasks())
        .map(ApiResponse::success)
        .or_else(|e| Ok(ApiResponse::error_msg("LIST_ERROR", &e)))
}

/// Lists recent import tasks within the last `days` (default: 7).
#[tauri::command]
pub async fn list_recent_import_tasks(
    state: State<'_, ImportTasksState>,
    days: Option<i64>,
) -> Result<ApiResponse<Vec<PersistedImportTask>>, String> {
    let lookback_days = days.unwrap_or(7);
    state
        .with_repo(|repo| repo.list_tasks_since_days(lookback_days))
        .map(ApiResponse::success)
        .or_else(|e| Ok(ApiResponse::error_msg("LIST_ERROR", &e)))
}

/// Lists tasks that can be retried (error or interrupted status).
#[tauri::command]
pub async fn list_retryable_tasks(
    state: State<'_, ImportTasksState>,
) -> Result<ApiResponse<Vec<PersistedImportTask>>, String> {
    state
        .with_repo(|repo| {
            let mut tasks = repo.list_tasks_by_status(ImportTaskStatus::Error)?;
            tasks.extend(repo.list_tasks_by_status(ImportTaskStatus::Interrupted)?);
            Ok(tasks)
        })
        .map(ApiResponse::success)
        .or_else(|e| Ok(ApiResponse::error_msg("LIST_ERROR", &e)))
}

// =============================================================================
// MANAGEMENT COMMANDS
// =============================================================================

/// Creates missing product records from completed import tasks.
///
/// Useful to migrate existing users who have import history but an empty products database.
#[tauri::command]
pub async fn backfill_products_from_import_tasks(
    state: State<'_, ImportTasksState>,
    products_db_state: State<'_, DbState>,
) -> Result<ApiResponse<usize>, String> {
    info!("Backfilling products from import tasks");

    let tasks = match state.with_repo(|repo| repo.list_tasks_by_status(ImportTaskStatus::Done)) {
        Ok(tasks) => tasks,
        Err(e) => return Ok(ApiResponse::error_msg("BACKFILL_ERROR", &e)),
    };

    let mut created = 0usize;
    for task in tasks {
        match upsert_product_from_task(&products_db_state, &task) {
            Ok(true) => created += 1,
            Ok(false) => {}
            Err(e) => warn!("Failed to backfill product for task {}: {}", task.id, e),
        }
    }

    Ok(ApiResponse::success(created))
}

/// Prepares a task for retry by resetting its status to pending.
#[tauri::command]
pub async fn prepare_task_retry(
    state: State<'_, ImportTasksState>,
    id: String,
) -> Result<ApiResponse<bool>, String> {
    info!("Preparing task for retry: {}", id);

    // Validate source path still exists
    let task_opt = state
        .with_repo(|repo| repo.get_task(&id))
        .map_err(|e| e.to_string())?;

    match task_opt {
        Some(task) => {
            let source_path = Path::new(&task.source_path);
            if !source_path.exists() {
                return Ok(ApiResponse::error(AppError::NotFound(
                    source_path.to_path_buf(),
                )));
            }
        }
        None => {
            return Ok(ApiResponse::error(AppError::NotFound(PathBuf::from(id))));
        }
    }

    state
        .with_repo(|repo| {
            repo.prepare_retry(&id)?;
            Ok(true)
        })
        .map(ApiResponse::success)
        .or_else(|e| Ok(ApiResponse::error_msg("RETRY_ERROR", &e)))
}

/// Deletes a task from the database.
#[tauri::command]
pub async fn delete_import_task(
    state: State<'_, ImportTasksState>,
    id: String,
) -> Result<ApiResponse<bool>, String> {
    info!("Deleting import task: {}", id);

    state
        .with_repo(|repo| {
            repo.delete_task(&id)?;
            Ok(true)
        })
        .map(ApiResponse::success)
        .or_else(|e| Ok(ApiResponse::error_msg("DELETE_ERROR", &e)))
}

/// Deletes tasks older than the specified number of days.
#[tauri::command]
pub async fn cleanup_old_import_tasks(
    state: State<'_, ImportTasksState>,
    days: i64,
) -> Result<ApiResponse<usize>, String> {
    info!("Cleaning up tasks older than {} days", days);

    state
        .with_repo(|repo| repo.cleanup_old_tasks(days))
        .map(ApiResponse::success)
        .or_else(|e| Ok(ApiResponse::error_msg("CLEANUP_ERROR", &e)))
}

/// Deletes all completed tasks (done or error status).
#[tauri::command]
pub async fn clear_completed_import_tasks(
    state: State<'_, ImportTasksState>,
) -> Result<ApiResponse<usize>, String> {
    info!("Clearing all completed import tasks");

    state
        .with_repo(|repo| repo.delete_completed_tasks())
        .map(ApiResponse::success)
        .or_else(|e| Ok(ApiResponse::error_msg("CLEAR_ERROR", &e)))
}
